### Basic Arithmetic

In [1]:
# Addition
2+1

3

In [2]:
# Subtraction
2-1

1

In [3]:
# Multiplication
2*2

4

In [4]:
# Division
3/2

1.5

In [5]:
# Floor Division
7//4

1

**So what if we just want the remainder after division?**

In [6]:
# Modulo
7%4

3

In [7]:
# Powers
2**3

8

In [8]:
# Can also do roots this way
4**0.5

2.0

In [9]:
# Order of Operations followed in Python
2 + 10 * 10 + 3

105

In [10]:
# Can use parentheses to specify orders
(2+10) * (10+3)

156

## Variable Assignments

In [11]:
# Let's create an object called "a" and assign it the number 5
a = 5

In [12]:
# Adding the objects
a+a

10

In [13]:
# Reassignment
a = 10

In [14]:
# Check
a

10

In [15]:
# Use A to redefine A
a = a + a

In [16]:
# Check 
a

20

In [17]:
# Use object names to keep better track of what's going on in your code!
my_income = 100

tax_rate = 0.1

my_taxes = my_income*tax_rate

In [18]:
# Show my taxes!
my_taxes

10.0

In [19]:
type(a)

int

Common data types include:
* **int** (for integer)
* **float**
* **str** (for string)
* **list**
* **tuple**
* **dict** (for dictionary)
* **set**
* **bool** (for Boolean True/False)

## Creating a String
To create a string in Python you need to use either single quotes or double quotes. For example:

In [20]:
# Single word
'hello'

'hello'

In [21]:
# Entire phrase 
'This is also a string'

'This is also a string'

In [22]:
# We can also use double quote
"String built with double quotes"

'String built with double quotes'

In [23]:
# Be careful with quotes!
' I'm using single quotes, but this will create an error'

SyntaxError: invalid syntax (<ipython-input-23-da9a34b3dc31>, line 2)

The reason for the error above is because the single quote in <code>I'm</code> stopped the string. You can use combinations of double and single quotes to get the complete statement.

In [24]:
"Now I'm ready to use the single quotes inside a string!"

"Now I'm ready to use the single quotes inside a string!"

In [25]:
# Note that we can't output multiple strings this way
'Hello World 1'
'Hello World 2'

'Hello World 2'

In [26]:
print('Hello World 1')
print('Hello World 2')
print('Use \n to print a new line')
print('\n')
print('See what I mean?')

Hello World 1
Hello World 2
Use 
 to print a new line


See what I mean?


In [27]:
len('Hello World')

11

## String Indexing


In [28]:
# Assign s as a string
s = 'Hello World'

In [29]:
#Check
s

'Hello World'

In [30]:
# Print the object
print(s) 

Hello World


In [31]:
# Show first element (in this case a letter)
s[0]

'H'

In [32]:
s[1]

'e'

In [33]:
s[2]

'l'

We can use a <code>:</code> to perform *slicing* which grabs everything up to a designated point. For example:

In [34]:
# Grab everything past the first term all the way to the length of s which is len(s)
s[1:]

'ello World'

In [35]:
# Note that there is no change to the original s
s

'Hello World'

In [36]:
# Grab everything UP TO the 3rd index
s[:3]

'Hel'

Note the above slicing. Here we're telling Python to grab everything from 0 up to 3. It doesn't include the 3rd index. You'll notice this a lot in Python, where statements and are usually in the context of "up to, but not including".

In [37]:
#Everything
s[:]

'Hello World'

We can also use negative indexing to go backwards.

In [38]:
# Last letter (one index behind 0 so it loops back around)
s[-1]

'd'

In [39]:
# Grab everything but the last letter
s[:-1]

'Hello Worl'

We can also use index and slice notation to grab elements of a sequence by a specified step size (the default is 1). For instance we can use two colons in a row and then a number specifying the frequency to grab elements. For example:

In [40]:
# Grab everything, but go in steps size of 1
s[::1]

'Hello World'

In [41]:
# Grab everything, but go in step sizes of 2
s[::2]

'HloWrd'

In [42]:
# We can use this to print a string backwards
s[::-1]

'dlroW olleH'

## String Properties
It's important to note that strings have an important property known as *immutability*. This means that once a string is created, the elements within it can not be changed or replaced. For example:

In [43]:
s

'Hello World'

In [44]:
# Let's try to change the first letter to 'x'
s[0] = 'x'

TypeError: 'str' object does not support item assignment

Notice how the error tells us directly what we can't do, change the item assignment!

Something we *can* do is concatenate strings!

In [45]:
s

'Hello World'

In [46]:
# Concatenate strings!
s + ' concatenate me!'

'Hello World concatenate me!'

In [47]:
# We can reassign s completely though!
s = s + ' concatenate me!'

In [48]:
print(s)

Hello World concatenate me!


In [49]:
s

'Hello World concatenate me!'

We can use the multiplication symbol to create repetition!

In [50]:
letter = 'z'

In [51]:
letter*10

'zzzzzzzzzz'

## Basic Built-in String methods

Objects in Python usually have built-in methods. 

Methods are in the form:  object.method(parameters)


In [52]:
s

'Hello World concatenate me!'

In [53]:
# Upper Case a string
s.upper()

'HELLO WORLD CONCATENATE ME!'

In [54]:
# Lower case
s.lower()

'hello world concatenate me!'

In [55]:
# Split a string by blank space (this is the default)
s.split()

['Hello', 'World', 'concatenate', 'me!']

In [56]:
# Split by a specific element (doesn't include the element that was split on)
s.split('W')

['Hello ', 'orld concatenate me!']

# String Formatting

String formatting lets you inject items into a string rather than trying to chain items together using commas or string concatenation. As a quick comparison, consider:

    player = 'Thomas'
    points = 33
    
    'Last night, '+player+' scored '+str(points)+' points.'  # concatenation
    
    f'Last night, {player} scored {points} points.'          # string formatting


There are three ways to perform string formatting.
* The oldest method involves placeholders using the modulo `%` character.
* An improved technique uses the `.format()` string method.

## Formatting with placeholders
You can use <code>%s</code> to inject strings into your print statements. The modulo `%` is referred to as a "string formatting operator".

In [57]:
print("I'm going to inject %s here." %'something')

I'm going to inject something here.


You can pass multiple items by placing them inside a tuple after the `%` operator.

In [58]:
print("I'm going to inject %s text here, and %s text here." %('some','more'))

I'm going to inject some text here, and more text here.


You can also pass variable names:

In [59]:
x, y = 'some', 'more'
print("I'm going to inject %s text here, and %s text here."%(x,y))

I'm going to inject some text here, and more text here.


### Format conversion methods.
It should be noted that two methods <code>%s</code> and <code>%r</code> convert any python object to a string using two separate methods: `str()` and `repr()`. We will learn more about these functions later on in the course, but you should note that `%r` and `repr()` deliver the *string representation* of the object, including quotation marks and any escape characters.

In [60]:
print('He said his name was %s.' %'Fred')
print('He said his name was %r.' %'Fred')

He said his name was Fred.
He said his name was 'Fred'.


As another example, `\t` inserts a tab into a string.

In [61]:
print('I once caught a fish %s.' %'this \tbig')
print('I once caught a fish %r.' %'this \tbig')

I once caught a fish this 	big.
I once caught a fish 'this \tbig'.


The `%s` operator converts whatever it sees into a string, including integers and floats. The `%d` operator converts numbers to integers first, without rounding. Note the difference below:

In [62]:
print('I wrote %s programs today.' %3.75)
print('I wrote %d programs today.' %3.75)   

I wrote 3.75 programs today.
I wrote 3 programs today.


### Padding and Precision of Floating Point Numbers
Floating point numbers use the format <code>%5.2f</code>. Here, <code>5</code> would be the minimum number of characters the string should contain; these may be padded with whitespace if the entire number does not have this many digits. Next to this, <code>.2f</code> stands for how many numbers to show past the decimal point. Let's see some examples:

In [63]:
print('Floating point numbers: %5.2f' %(13.144))

Floating point numbers: 13.14


In [64]:
print('Floating point numbers: %1.0f' %(13.144))

Floating point numbers: 13


In [65]:
print('Floating point numbers: %1.5f' %(13.144))

Floating point numbers: 13.14400


In [66]:
print('Floating point numbers: %10.2f' %(13.144))

Floating point numbers:      13.14


In [67]:
print('Floating point numbers: %25.2f' %(13.144))

Floating point numbers:                     13.14


### Multiple Formatting

In [68]:
print('First: %s, Second: %5.2f, Third: %r' %('hi!',3.1415,'bye!'))

First: hi!, Second:  3.14, Third: 'bye!'


## Formatting with the `.format()` method
A better way to format objects into your strings for print statements is with the string `.format()` method. The syntax is:

    'String here {} then also {}'.format('something1','something2')
    
For example:

In [69]:
print('This is a string with an {}'.format('insert'))

This is a string with an insert


### The .format() method has several advantages over the %s placeholder method:

#### 1. Inserted objects can be called by index position:

In [70]:
print('The {2} {1} {0}'.format('fox','brown','quick'))

The quick brown fox


#### 2. Inserted objects can be assigned keywords:

In [71]:
print('First Object: {a}, Second Object: {b}, Third Object: {c}'.format(a=1,b='Two',c=12.3))

First Object: 1, Second Object: Two, Third Object: 12.3


#### 3. Inserted objects can be reused, avoiding duplication:

In [72]:
print('A %s saved is a %s earned.' %('penny','penny'))
# vs.
print('A {p} saved is a {p} earned.'.format(p='penny'))

A penny saved is a penny earned.
A penny saved is a penny earned.


### Alignment, padding and precision with `.format()`
Within the curly braces you can assign field lengths, left/right alignments, rounding parameters and more

In [73]:
print('{0:8} | {1:9}'.format('Fruit', 'Quantity'))
print('{0:8} | {1:9}'.format('Apples', 3.))
print('{0:8} | {1:9}'.format('Oranges', 10))

Fruit    | Quantity 
Apples   |       3.0
Oranges  |        10


By default, `.format()` aligns text to the left, numbers to the right. You can pass an optional `<`,`^`, or `>` to set a left, center or right alignment:

In [74]:
print('{0:<8} | {1:^8} | {2:>8}'.format('Left','Center','Right'))
print('{0:<8} | {1:^8} | {2:>8}'.format(11,22,33))

Left     |  Center  |    Right
11       |    22    |       33


You can precede the aligment operator with a padding character

In [75]:
print('{0:=<8} | {1:-^8} | {2:.>8}'.format('Left','Center','Right'))
print('{0:=<8} | {1:-^8} | {2:.>8}'.format(11,22,33))

Left==== | -Center- | ...Right
11====== | ---22--- | ......33


Field widths and float precision are handled in a way similar to placeholders. The following two print statements are equivalent:

In [76]:
print('This is my ten-character, two-decimal number:%10.2f' %13.579)
print('This is my ten-character, two-decimal number:{0:10.2f}'.format(13.579))

This is my ten-character, two-decimal number:     13.58
This is my ten-character, two-decimal number:     13.58


Note that there are 5 spaces following the colon, and 5 characters taken up by 13.58, for a total of ten characters.
